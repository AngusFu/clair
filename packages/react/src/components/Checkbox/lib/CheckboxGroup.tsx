import React, { useState } from "react";
import classNames from "classnames";
import PropTypes from "prop-types";

import {
  CheckboxGroupContextType,
  CheckboxGroupProps,
  CheckboxGroupHandleValue,
  UpdateGroupValueType,
  CheckboxGroupValueType
} from "./types";

export const CheckboxGroupContext = React.createContext<
  CheckboxGroupContextType
>({});

export const CheckboxGroup: React.FC<CheckboxGroupProps> = props => {
  const {
    className,
    name,
    style,
    children,
    value: valueProp,
    defaultValue: defaultValueProp,
    onChange
  } = props;
  const isControlled = "value" in props;
  const defaultValue = defaultValueProp || [];
  const [uncontrolledValue, updateUncontrolledValue] = useState(defaultValue);

  const value2Use: CheckboxGroupValueType = isControlled
    ? (valueProp as CheckboxGroupValueType)
    : uncontrolledValue;

  const registerValue: CheckboxGroupHandleValue = value => {
    if (value2Use.includes(value)) return;
    const newValue = value2Use.concat(value);
    if (!isControlled) {
      updateUncontrolledValue(newValue);
    }
    onChange &&
      onChange({
        target: {
          value: newValue
        }
      });
  };

  const cancelValue: CheckboxGroupHandleValue = value => {
    const newValue = value2Use.filter(val => val !== value);
    if (newValue.length === value2Use.length) return;
    if (!isControlled) {
      updateUncontrolledValue(newValue);
    }
    onChange &&
      onChange({
        target: {
          value: newValue
        }
      });
  };

  const updateGroupValue: UpdateGroupValueType = (isChecked, value) =>
    isChecked ? registerValue(value) : cancelValue(value);

  return (
    <CheckboxGroupContext.Provider
      value={{
        groupValue: value2Use,
        name,
        updateGroupValue
      }}
    >
      <div className={classNames("c-checkbox-group", className)} style={style}>
        {children}
      </div>
    </CheckboxGroupContext.Provider>
  );
};

CheckboxGroup.propTypes = {
  className: PropTypes.string,
  name: PropTypes.string,
  style: PropTypes.object,
  value: PropTypes.arrayOf(PropTypes.string),
  onChange: PropTypes.func,
  defaultValue: PropTypes.arrayOf(PropTypes.string),
  children: PropTypes.node
} as PropTypes.ValidationMap<CheckboxGroupProps>;
CheckboxGroup.displayName = "CheckboxGroup";
