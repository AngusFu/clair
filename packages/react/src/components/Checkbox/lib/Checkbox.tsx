import React, { useState, useRef, useEffect, useContext } from "react";
import classNames from "classnames";
import PropTypes from "prop-types";

import {
  UseCheckedType,
  UseNameType,
  CheckboxProps,
  CheckboxEvent
} from "./types";
import { CheckboxGroupContext } from "./CheckboxGroup";
import { AutoIncreasingCounter } from "@clair/helpers";

const sizes = ["small", "normal", "large"] as const;
const propSizes = PropTypes.oneOf(sizes);
const autoIncrCounter = new AutoIncreasingCounter();

const useChecked: UseCheckedType = (props, groupContext) => {
  const { checked, defaultChecked, value, onChange } = props;
  const { groupValue, updateGroupValue } = groupContext;

  const [uncontrolledChecked, updateUncontrolledChecked] = useState(
    defaultChecked
  );
  const isControlled = "checked" in props;
  const isGroupControlled = Array.isArray(groupValue);

  let checked2Use = Boolean(isControlled ? checked : uncontrolledChecked);
  if (isGroupControlled) {
    if (!value) {
      throw new Error(
        `Invalid props 'value'! Value must be set when controlled by CheckboxGroup.`
      );
    }
    checked2Use = (groupValue as string[]).includes(value);
  }

  const updateChecked = (e: CheckboxEvent) => {
    const {
      target: { checked: newCheckedStatus }
    } = e;
    onChange?.(e);
    if (isGroupControlled) {
      return (
        updateGroupValue && updateGroupValue(newCheckedStatus, value as string)
      );
    }
    !isControlled && updateUncontrolledChecked(newCheckedStatus);
  };

  return [checked2Use, updateChecked];
};

const useName: UseNameType = ({ name = "" }) => {
  const { name: groupName } = useContext(CheckboxGroupContext);
  return groupName || name;
};

export const Checkbox: React.FC<CheckboxProps> = props => {
  const {
    className,
    disabled,
    indeterminate,
    style,
    size,
    value,
    children
  } = props;
  const curRef = useRef<HTMLInputElement>(null);
  const { current: inputId } = useRef(
    `c-checkbox__input-${autoIncrCounter.next()}`
  );
  const { current: labelId } = useRef(
    `c-checkbox__label-${autoIncrCounter.next()}`
  );

  const groupContext = useContext(CheckboxGroupContext);
  const [checked2Use, updateChecked] = useChecked(props, groupContext);
  const name = useName(props);

  useEffect(() => {
    (curRef.current as HTMLInputElement).indeterminate = Boolean(indeterminate);
  }, [indeterminate]);

  return (
    <label
      className={classNames("c-checkbox", className, {
        [`c-checkbox--${size}`]: sizes.some(curSize => size === curSize)
      })}
      style={style}
      htmlFor={inputId}
    >
      <input
        id={inputId}
        aria-labelledby={labelId}
        type="checkbox"
        ref={curRef}
        disabled={disabled}
        checked={Boolean(checked2Use)}
        name={name}
        value={value}
        onChange={updateChecked}
      />
      <span className="c-checkbox__box" />
      <span className="c-checkbox__label" id={labelId}>
        {children}
      </span>
    </label>
  );
};

Checkbox.propTypes = {
  className: PropTypes.string,
  name: PropTypes.string,
  style: PropTypes.object,
  checked: PropTypes.bool,
  defaultChecked: PropTypes.bool,
  size: propSizes,
  indeterminate: PropTypes.bool,
  disabled: PropTypes.bool,
  value: PropTypes.string,
  children: PropTypes.node,
  onChange: PropTypes.func
};
Checkbox.displayName = "Checkbox";
