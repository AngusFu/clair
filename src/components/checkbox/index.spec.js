import { mount } from '@vue/test-utils'
import Checkbox from './index.vue'
import CheckboxGroup from './checkbox-group.vue'

it('should render basic checkbox correctly', () => {
  const wrapper = mount(Checkbox, {
    propsData: {
      value: true
    }
  })

  expect(wrapper.vm.checked).toBe(true)
  wrapper.trigger('click')
  expect(wrapper.vm.checked).toBe(false)

  expect(wrapper).toMatchSnapshot()
})

it('should render checkbox group correctly', () => {
  const wrapper = mount(CheckboxGroup, {
    propsData: {
      options: [
        { value: 1, label: '选项1' },
        { value: 2, label: '选项2' },
        { value: 3, label: '选项3', disabled: true },
        { value: 4, label: '选项3', disabled: true }
      ],
      value: [4]
    },
    listeners: {
      change (val) {
        wrapper.setProps({ value: val })
      }
    }
  })

  expect(wrapper).toMatchSnapshot()

  const checkboxes = wrapper.findAll(Checkbox)

  checkboxes.at(0).trigger('click')
  expect(wrapper.vm.value).toEqual([4, 1])

  checkboxes.at(1).trigger('click')
  expect(wrapper.vm.value).toEqual([4, 1, 2])

  checkboxes.at(2).trigger('click')
  expect(wrapper.vm.value).toEqual([4, 1, 2])
})

it('should render checkbox group with custom children', () => {
  const wrapper = mount(CheckboxGroup, {
    propsData: {
      value: []
    },
    stubs: {
      'c-checkbox': Checkbox
    },
    slots: {
      default: [
        '<c-checkbox label="a">选项1</c-checkbox>',
        '<c-checkbox label="选项2"></c-checkbox>',
        '<c-checkbox label="选项3"></c-checkbox>'
      ]
    },
    listeners: {
      change (val) {
        wrapper.setProps({ value: val })
      }
    }
  })

  expect(wrapper).toMatchSnapshot()

  const checkboxes = wrapper.findAll(Checkbox)
  checkboxes.at(0).trigger('click')
  checkboxes.at(2).trigger('click')
  expect(wrapper.vm.value).toEqual(['a', '选项3'])
})

it('should render indeterminate state', () => {
  const wrapper = mount({
    components: {
      'c-checkbox': Checkbox,
      'c-checkbox-group': CheckboxGroup
    },
    template: `
<div>
  <c-checkbox
    v-model="allChecked"
    :indeterminate="indeterminate"
    label="全部选中"
    @change="onCheckAllChange"
  />
  <br/>
  <c-checkbox-group
    v-model="selected"
    :options="options"
    @change="onCheckedResultChange"
  />

  <p>你选择了 {{ selected }}</p>
</div>
`,
    data () {
      return {
        options: ['Node', 'Nginx', 'Vue'],
        selected: ['Node'],
        allChecked: false,
        indeterminate: true
      }
    },
    methods: {
      onCheckAllChange (e) {
        this.selected = !e ? [] : this.options
        this.indeterminate = false
      },
      onCheckedResultChange (e) {
        const total = this.options.length
        const checked = this.selected.length
        this.allChecked = e.length === total
        this.indeterminate = checked > 0 && checked < total
      }
    }
  })

  expect(wrapper).toMatchSnapshot()

  wrapper.findAll(Checkbox).at(2).trigger('click')
  wrapper.findAll(Checkbox).at(3).trigger('click')
  expect(wrapper.vm.indeterminate).toBe(false)
})

it('should works with validation', () => {
  const wrapper = mount(CheckboxGroup, {
    propsData: {
      value: [],
      options: [
        { value: 1, label: '香蕉' },
        { value: 2, label: '苹果' },
        { value: 3, label: '梨' },
        { value: 4, label: '芒果' },
        { value: 5, label: '木瓜' },
        { value: 6, label: '榴莲' }
      ],
      minItems: 2,
      maxItems: 4
    },
    listeners: {
      change (val) {
        wrapper.setProps({ value: val })
      }
    }
  })

  const checkboxes = wrapper.findAll(Checkbox)
  // select five items
  checkboxes.at(0).trigger('click')
  checkboxes.at(1).trigger('click')
  checkboxes.at(2).trigger('click')
  checkboxes.at(3).trigger('click')
  checkboxes.at(4).trigger('click')

  expect(wrapper).toMatchSnapshot()
  expect(wrapper.contains('.c-error-msg')).toEqual(true)
})
