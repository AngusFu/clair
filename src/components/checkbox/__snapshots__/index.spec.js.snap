// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render basic checkbox correctly 1`] = `<label class="c-checkbox"><input type="checkbox"><span class="c-checkbox__box"></span><span class="c-checkbox__label"></span></label>`;

exports[`should render checkbox group correctly 1`] = `
<div class="c-checkbox-group"><label class="c-checkbox"><input type="checkbox"><span class="c-checkbox__box"></span><span class="c-checkbox__label">选项1</span></label><label class="c-checkbox"><input type="checkbox"><span class="c-checkbox__box"></span><span class="c-checkbox__label">选项2</span></label><label class="c-checkbox"><input type="checkbox" disabled="disabled"><span class="c-checkbox__box"></span><span class="c-checkbox__label">选项3</span></label><label class="c-checkbox"><input type="checkbox" disabled="disabled"><span class="c-checkbox__box"></span><span class="c-checkbox__label">选项3</span></label>
  <!---->
</div>
`;

exports[`should render checkbox group with custom children 1`] = `
<div class="c-checkbox-group"><label class="c-checkbox"><input type="checkbox"><span class="c-checkbox__box"></span><span class="c-checkbox__label">选项1</span></label><label class="c-checkbox"><input type="checkbox"><span class="c-checkbox__box"></span><span class="c-checkbox__label">选项2</span></label><label class="c-checkbox"><input type="checkbox"><span class="c-checkbox__box"></span><span class="c-checkbox__label">选项3</span></label>
  <!---->
</div>
`;

exports[`should render indeterminate state 1`] = `
<div><label class="c-checkbox"><input type="checkbox"><span class="c-checkbox__box"></span><span class="c-checkbox__label">全部选中</span></label> <br>
  <div class="c-checkbox-group"><label class="c-checkbox"><input type="checkbox"><span class="c-checkbox__box"></span><span class="c-checkbox__label">Node</span></label><label class="c-checkbox"><input type="checkbox"><span class="c-checkbox__box"></span><span class="c-checkbox__label">Nginx</span></label><label class="c-checkbox"><input type="checkbox"><span class="c-checkbox__box"></span><span class="c-checkbox__label">Vue</span></label>
    <!---->
  </div>
  <p>你选择了 [
    "Node"
    ]</p>
</div>
`;

exports[`should works with validation 1`] = `<div class="c-checkbox-group"><label class="c-checkbox"><input type="checkbox"><span class="c-checkbox__box"></span><span class="c-checkbox__label">香蕉</span></label><label class="c-checkbox"><input type="checkbox"><span class="c-checkbox__box"></span><span class="c-checkbox__label">苹果</span></label><label class="c-checkbox"><input type="checkbox"><span class="c-checkbox__box"></span><span class="c-checkbox__label">梨</span></label><label class="c-checkbox"><input type="checkbox"><span class="c-checkbox__box"></span><span class="c-checkbox__label">芒果</span></label><label class="c-checkbox"><input type="checkbox"><span class="c-checkbox__box"></span><span class="c-checkbox__label">木瓜</span></label><label class="c-checkbox"><input type="checkbox"><span class="c-checkbox__box"></span><span class="c-checkbox__label">榴莲</span></label><em class="c-error-msg">最多可以选择4项</em></div>`;
